man/mergelist.Rd

```diff
@@ -0,0 +1,150 @@
+}
+\examples{
+l = list(
+  data.table(id1 = c(1:4,2:5), v1 = 1:8),
+  data.table(id1 = 2:3, v2 = 1:2),
+  data.table(id1 = 3:5, v3 = 1:3)
+)
+mergelist(l, on="id1")
+
+## using keys
+l = list(
+  data.table(id1 = c(1:4,2:5), v1 = 1:8),
+  data.table(id1 = 3:5, id2 = 1:3, v2 = 1:3, key="id1"),
+  data.table(id2 = 1:4, v3 = 4:1, key="id2")
+)
+mergelist(l)
+
+## select columns
+l = list(
+  data.table(id1 = c(1:4,2:5), v1 = 1:8, v2 = 8:1),
+  data.table(id1 = 3:5, v3 = 1:3, v4 = 3:1, v5 = 1L, key="id1")
+)
+mergelist(l, cols = list(NULL, c("v3","v5")))
+
+## detecting duplicates matches
+l = list(
+  data.table(id1=c(1:4,2:5), v1=1:8), ## dups in LHS are fine
+  data.table(id1=c(2:3,2L), v2=1:3),  ## dups in RHS
+  data.table(id1=3:5, v3=1:3)
+)
+#mergelist(l, on="id1") ## mult='error' and multiple matches during merge
+lapply(l[-1L], function(d) copy(d)[,cnt:=.N,by="id1"][cnt>1L])
+
+## denormalize star schema
+
+### make fact and two dimensions
+
+gt = state.x77[,"Population"]
+gt = data.table(state_id=seq_along(state.name), p=gt[state.name]/sum(gt), k=1L)
+tt = as.IDate(paste0(as.integer(time(uspop)),"-01-01"))
+tt = as.data.table(stats::approx(tt, c(uspop), tt[1L]:tt[length(tt)]))
+tt = tt[, .(date=as.IDate(x), time_id=seq_along(x), pop=y, k=1L)]
+fact = tt[gt, on="k", allow.cartesian=TRUE,
+          .(state_id=i.state_id, time_id=x.time_id, population = x.pop * i.p)]
+setkeyv(fact, c("state_id","time_id"))
+
+time = data.table(key = "time_id",
+  time_id = seq_along(tt$date), date = tt$date,
+  month_id = month(tt$date), month = month.name[month(tt$date)],
+  year_id = year(tt$date)-1789L, year = as.character(year(tt$date)),
+  week_id = week(tt$date), week = as.character(week(tt$date)),
+  weekday_id = wday(tt$date)-1L, weekday = weekdays(tt$date)
+)[weekday_id==0L, weekday_id:=7L][]
+
+geog = data.table(key = "state_id",
+  state_id = seq_along(state.abb), state_abb = state.abb, state_name = state.name,
+  division_id = as.integer(state.division),
+  division_name = as.character(state.division),
+  region_id = as.integer(state.region),
+  region_name = as.character(state.region)
+)
+
+X = list(fact, time, geog)
+ans = mergelist(X)
+
+rm(gt, tt, ans)
+
+## denormalize snowflake schema
+
+make.lvl = function(x, cols) {
+  stopifnot(is.data.table(x))
+  lvl = x[, unique(.SD), .SDcols=cols]
+  setkeyv(lvl, cols[1L])
+  setindexv(lvl, as.list(cols))
+}
+time = list(
+  date = make.lvl(time, c("time_id","date","year_id","month_id","week_id",
+                          "weekday_id")),
+  weekday = make.lvl(time, c("weekday_id","weekday")),
+  week = make.lvl(time, c("week_id","week")),
+  month = make.lvl(time, c("month_id","month")),
+  year = make.lvl(time, c("year_id","year"))
+)
+geog = list(
+  state = make.lvl(geog, c("state_id","state_abb","state_name","division_id")),
+  division = make.lvl(geog, c("division_id","division_name","region_id")),
+  region = make.lvl(geog, c("region_id","region_name"))
+)
```

Comment:
I think we should address such case better, maybe by better optimizations for right outer join, to verify.
`m` stands for merging pair
```r
## current model
m(         ## division.region_id already lost
  m(       ## state.division_id already lost
    m(     ## fact.state_id used
      fact,
      state
    ),
    division
  ),
  region
)

## what should be possible
m(         ## fact.state_id index used
  fact,
  m(       ## state.division_id index used
    state,
    m(     ## division.region_id index used
      division,
      region
    )
  )
)
```

Reply:
Actually, in left join, the index was there, so still was used, but the key was lost on one side. Right outer join seems to address that better, by retaining keys, which is a result of a sequence in which merges are applied, as presented above. Below is verbose output of left and right joins, where we can see the sequence of joins. Few `forder` calls will be eliminated once #4386 will be merged.
```
                                     left                                    right
 i.state_id has same type (integer) as x. i.region_id has same type (integer) as x
      on= matches existing key, using key      on= matches existing key, using key
                      Starting bmerge ...                      Starting bmerge ...
 forder.c received 1000000 rows and 2 col forder.c received 10000 rows and 2 colum
 bmerge done in 0.184s elapsed (0.159s cp bmerge done in 0.001s elapsed (0.001s cp
                   cbindlist: took 0.000s                   cbindlist: took 0.000s
 i.division_id has same type (integer) as i.division_id has same type (integer) as
      on= matches existing key, using key      on= matches existing key, using key
                      Starting bmerge ...                      Starting bmerge ...
 forder.c received 1000000 rows and 3 col forder.c received 100000 rows and 2 colu
 bmerge done in 0.134s elapsed (0.129s cp bmerge done in 0.009s elapsed (0.005s cp
                   cbindlist: took 0.000s                   cbindlist: took 0.000s
 i.region_id has same type (integer) as x i.state_id has same type (integer) as x.
      on= matches existing key, using key      on= matches existing key, using key
                      Starting bmerge ...                      Starting bmerge ...
 forder.c received 1000000 rows and 4 col forder.c received 1000000 rows and 2 col
 bmerge done in 0.106s elapsed (0.099s cp bmerge done in 0.169s elapsed (0.168s cp
                   cbindlist: took 0.000s                   cbindlist: took 0.000s
 mergelist: merging 4 tables, took 0.570s mergelist: merging 4 tables, took 0.215s
```
```r
set.seed(108)
N = 1e6L

region = data.table(region_id=seq_len(N/1000), key="region_id")
division = data.table(division_id=seq_len(N/100), region_id=sample(N/1000, N/100, TRUE), key="division_id")
setindexv(division, "region_id")
state = data.table(state_id=seq_len(N/10), division_id=sample(N/100, N/10, TRUE), key="state_id")
setindexv(state, "division_id")
fact = data.table(state_id=sample(N/10, N, TRUE), population=1)
setindexv(fact, "state_id")
l = list(fact, state, division, region)
options(datatable.verbose=TRUE)
out1 = capture.output(ans1<-mergelist(l, how="left"))
out2 = capture.output(ans2<-mergelist(rev(l), how="right"))
options(datatable.verbose=FALSE)
```

Reply:
some more timings on that matter https://gist.github.com/jangorecki/ef1bb100cdd46f77f84c3aeda56fbe41
```
> ## medium cardinality
> sapply(l, nrow)
[1] 100000000  10000000   1000000    100000
> system.time(ans1<-mergelist(l, how="left"))
   user  system elapsed 
 66.078  10.530  19.845 
> system.time(ans2<-mergelist(rev(l), how="right"))
   user  system elapsed 
 29.692   4.947   9.334 
> 
> ## high cardinality
> sapply(l, nrow)
[1] 100000000  50000000  25000000  12500000
> system.time(ans1<-mergelist(l, how="left"))
   user  system elapsed 
 95.045  13.896  32.975 
> system.time(ans2<-mergelist(rev(l), how="right"))
   user  system elapsed 
 69.442  10.305  22.856 
> 
> ## low cardinality
> sapply(l, nrow)
[1] 100000000   1000000     10000       100
> system.time(ans1<-mergelist(l, how="left"))
   user  system elapsed 
 42.330  11.342  16.467 
> system.time(ans2<-mergelist(rev(l), how="right"))
   user  system elapsed 
 23.495   5.346   7.861 
```